File: FakturusTrack.Android/app/src/main/java/com/fakturus/track/features/settings/SettingsViewModel.kt
package com.fakturus.track.features.settings

import android.content.Context
import android.util.Log
import androidx.datastore.preferences.core.booleanPreferencesKey
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.fakturus.track.BuildConfig
import com.fakturus.track.models.AppDatabase
import com.fakturus.track.models.UserSettingsEntity
import com.fakturus.track.models.UserSettingsHistoryEntryDTO
import com.fakturus.track.R
import com.fakturus.track.services.subscription.BillingManager
import com.fakturus.track.services.subscription.SubscriptionManager
import com.fakturus.track.services.subscription.Tier
import com.fakturus.track.services.api.APIClient
import com.fakturus.track.services.sync.SyncEngine
import com.fakturus.track.settingsDataStore
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.LocalDate

class SettingsViewModel(
    private val database: AppDatabase,
    private val syncEngine: SyncEngine? = null,
    private val apiClient: APIClient? = null,
    private val context: Context? = null,
    private val billingManager: BillingManager? = null,
    private val subscriptionManager: SubscriptionManager? = null
) : ViewModel() {

    private val settingsDao = database.userSettingsDao()

    val settings: StateFlow<UserSettingsEntity?> = settingsDao.getSettings()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), null)

    private val _isSaving = MutableStateFlow(false)
    val isSaving: StateFlow<Boolean> = _isSaving.asStateFlow()

    // Appearance from DataStore
    val appearance: StateFlow<String> = context?.let { ctx ->
        ctx.settingsDataStore.data
            .map { prefs -> prefs[stringPreferencesKey("appearance")] ?: "system" }
            .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), "system")
    } ?: MutableStateFlow("system").asStateFlow()

    // Notifications from DataStore
    val notificationsEnabled: StateFlow<Boolean> = context?.let { ctx ->
        ctx.settingsDataStore.data
            .map { prefs -> prefs[booleanPreferencesKey("notifications")] ?: true }
            .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), true)
    } ?: MutableStateFlow(true).asStateFlow()

    val appVersion: String
        get() = "${BuildConfig.VERSION_NAME} (${BuildConfig.VERSION_CODE})"

    // Restore purchases
    private val _isRestoringPurchases = MutableStateFlow(false)
    val isRestoringPurchases: StateFlow<Boolean> = _isRestoringPurchases.asStateFlow()

    private val _restoreResult = MutableStateFlow<String?>(null)
    val restoreResult: StateFlow<String?> = _restoreResult.asStateFlow()

    fun restorePurchases() {
        viewModelScope.launch {
            _isRestoringPurchases.value = true
            try {
                billingManager?.queryExistingPurchases()
                val tier = subscriptionManager?.tier?.value ?: Tier.FREE
                _restoreResult.value = if (tier > Tier.FREE) {
                    context?.getString(R.string.restore_success, tier.name)
                } else {
                    context?.getString(R.string.restore_no_subscription)
                }
            } catch (e: Exception) {
                _restoreResult.value = context?.getString(R.string.restore_error)
            } finally {
                _isRestoringPurchases.value = false
            }
        }
    }

    private var debounceJob: Job? = null

    /**
     * Effective date for pending changes to historized fields (workDays /
     * workHoursPerWeek). Default = today. The settings UI exposes a date
     * picker so the user can backdate corrections.
     */
    private val _effectiveDate = MutableStateFlow(LocalDate.now())
    val effectiveDate: StateFlow<LocalDate> = _effectiveDate.asStateFlow()

    fun updateEffectiveDate(date: LocalDate) {
        _effectiveDate.value = date
        // Re-stage so the saved entity carries the new effective date.
        viewModelScope.launch {
            val current = settings.value ?: return@launch
            settingsDao.upsert(
                current.copy(
                    pendingEffectiveDate = date.toString(),
                    updatedAt = Instant.now().toString(),
                    isPendingSync = true,
                    isSynced = false
                )
            )
            syncEngine?.syncAll()
        }
    }

    /**
     * Latched: stays true while a historized change is staged but not yet
     * acknowledged by the server, so the "Gültig ab"-picker doesn't flicker
     * during the debounce/save round-trip.
     */
    private val _hasUnsyncedHistorizedChange = MutableStateFlow(false)
    val hasUnsyncedHistorizedChange: StateFlow<Boolean> = _hasUnsyncedHistorizedChange.asStateFlow()

    // History
    private val _settingsHistory = MutableStateFlow<List<UserSettingsHistoryEntryDTO>>(emptyList())
    val settingsHistory: StateFlow<List<UserSettingsHistoryEntryDTO>> = _settingsHistory.asStateFlow()
    private val _isLoadingHistory = MutableStateFlow(false)
    val isLoadingHistory: StateFlow<Boolean> = _isLoadingHistory.asStateFlow()
    private val _historyError = MutableStateFlow<String?>(null)
    val historyError: StateFlow<String?> = _historyError.asStateFlow()

    fun loadSettingsHistory() {
        val client = apiClient ?: return
        viewModelScope.launch {
            _isLoadingHistory.value = true
            _historyError.value = null
            try {
                _settingsHistory.value = client.getUserSettingsHistory()
            } catch (e: Exception) {
                Log.e("SettingsViewModel", "Failed to load history", e)
                _historyError.value = "Verlauf konnte nicht geladen werden"
            } finally {
                _isLoadingHistory.value = false
            }
        }
    }

    fun updateWorkHoursPerWeek(hours: Double) {
        updateSettings(historizedFieldChanged = true) { it.copy(workHoursPerWeek = hours) }
    }

    fun updateVacationDaysPerYear(days: Int) {
        updateSettings { it.copy(vacationDaysPerYear = days) }
    }

    fun updateWorkDays(workDays: Int) {
        updateSettings(historizedFieldChanged = true) { it.copy(workDays = workDays) }
    }

    fun updateBundesland(bundesland: String) {
        updateSettings { it.copy(bundesland = bundesland) }
    }

    fun updateCalendarUrl(url: String?) {
        updateSettings { it.copy(calendarUrl = url) }
    }

    fun updatePersonalNumber(number: String) {
        updateSettings { it.copy(personalNumber = number.ifBlank { null }) }
    }

    fun setAppearance(value: String) {
        val ctx = context ?: return
        viewModelScope.launch {
            ctx.settingsDataStore.edit { prefs ->
                prefs[stringPreferencesKey("appearance")] = value
            }
        }
    }

    fun setNotifications(enabled: Boolean) {
        val ctx = context ?: return
        viewModelScope.launch {
            ctx.settingsDataStore.edit { prefs ->
                prefs[booleanPreferencesKey("notifications")] = enabled
            }
        }
    }

    private fun updateSettings(
        historizedFieldChanged: Boolean = false,
        transform: (UserSettingsEntity) -> UserSettingsEntity
    ) {
        val current = settings.value ?: UserSettingsEntity(userId = "")
        val pendingEffectiveDate = if (historizedFieldChanged) {
            _effectiveDate.value.toString()
        } else {
            current.pendingEffectiveDate
        }
        val updated = transform(current).copy(
            updatedAt = Instant.now().toString(),
            isPendingSync = true,
            isSynced = false,
            pendingEffectiveDate = pendingEffectiveDate
        )

        // Latch picker visibility synchronously so it doesn't disappear during
        // the 500ms debounce window.
        if (historizedFieldChanged) {
            _hasUnsyncedHistorizedChange.value = true
        }

        debounceJob?.cancel()
        debounceJob = viewModelScope.launch {
            _isSaving.value = true
            delay(500) // Debounce 500ms
            try {
                settingsDao.upsert(updated)
                syncEngine?.syncAll()
                // After sync, re-read entity. If pendingEffectiveDate is still
                // set (offline / failure), keep the picker visible.
                val saved = settingsDao.getSettingsOnce()
                _hasUnsyncedHistorizedChange.value = saved?.pendingEffectiveDate != null
            } catch (e: Exception) {
                Log.e("SettingsViewModel", "Failed to save settings", e)
            } finally {
                _isSaving.value = false
            }
        }
    }

    fun initializeSettingsIfNeeded() {
        viewModelScope.launch {
            if (settings.value == null) {
                settingsDao.upsert(UserSettingsEntity(userId = ""))
            }
        }
    }

    fun deleteAccount(onSuccess: () -> Unit) {
        viewModelScope.launch {
            try {
                apiClient?.delete("/api/account")
                onSuccess()
            } catch (e: Exception) {
                Log.e("SettingsVM", "Account deletion failed", e)
            }
        }
    }
}

File: FakturusTrack.iOS/FakturusTrack/Features/Settings/SettingsViewModel.swift
import SwiftData
import SwiftUI
import Foundation
import Combine

@Observable @MainActor
final class SettingsViewModel {
    // MARK: - State

    var workHoursPerWeek: Double = 40.0
    var vacationDaysPerYear: Int = 30
    var workDays: Int = 31
    var bundesland: String = "NW"
    var schoolHolidays: [SchoolHolidayPeriod] = []
    var userName: String = ""
    var userEmail: String = ""
    var error: String?
    var isSaving = false

    /// Effective date for changes to `workDays` / `workHoursPerWeek`. Default
    /// is today. Picker lets the user backdate when correcting earlier entries.
    var effectiveDate: Date = Calendar.current.startOfDay(for: Date())

    /// True while a historized change is staged but not yet acknowledged by
    /// the server. Drives whether the "Gültig ab"-picker is visible. Stays on
    /// across the debounced save so the picker doesn't flicker; only flips
    /// back to false once the local model has its `pendingEffectiveDate`
    /// cleared (i.e. the server confirmed the change).
    var hasUnsyncedHistorizedChange: Bool = false

    /// Snapshot of the values at load time so we know whether the historized
    /// fields actually changed (and only then send `effectiveDate`).
    private var initialWorkDays: Int = 31
    private var initialWorkHoursPerWeek: Double = 40.0

    var workDaysOrHoursChanged: Bool {
        workDays != initialWorkDays || workHoursPerWeek != initialWorkHoursPerWeek
    }

    // MARK: - History (Stage 2)

    var settingsHistory: [UserSettingsHistoryEntryDTO] = []
    var isLoadingHistory: Bool = false
    var historyError: String?
    var apiClient: APIClient?

    // E10-S01: App settings
    @ObservationIgnored @AppStorage("notificationsEnabled") var notificationsEnabled: Bool = true

    var personalNumber: String = ""

    var appVersion: String {
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "?"
        let build = Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? "?"
        return "\(version) (\(build))"
    }

    // MARK: - Dependencies

    private let modelContext: ModelContext
    var syncEngine: SyncEngine?
    var authManager: AuthManager?

    // MARK: - Debounce

    private var debounceTask: Task<Void, Never>?
    private var personalNumberDebounceTask: Task<Void, Never>?

    init(modelContext: ModelContext, syncEngine: SyncEngine? = nil, authManager: AuthManager? = nil, apiClient: APIClient? = nil) {
        self.modelContext = modelContext
        self.syncEngine = syncEngine
        self.authManager = authManager
        self.apiClient = apiClient
        loadSettings()
        loadSchoolHolidays()
    }

    // MARK: - Load

    private func loadSettings() {
        let descriptor = FetchDescriptor<UserSettings>()
        guard let settings = try? modelContext.fetch(descriptor).first else { return }
        workHoursPerWeek = settings.workHoursPerWeek
        vacationDaysPerYear = settings.vacationDaysPerYear
        workDays = settings.workDays
        bundesland = settings.bundesland
        personalNumber = settings.personalNumber ?? ""
        initialWorkDays = settings.workDays
        initialWorkHoursPerWeek = settings.workHoursPerWeek
        // If we're entering the screen with a still-pending historized change
        // (e.g. last sync failed offline), keep the picker visible.
        hasUnsyncedHistorizedChange = settings.pendingEffectiveDate != nil
        effectiveDate = settings.pendingEffectiveDate ?? Calendar.current.startOfDay(for: Date())
    }

    func loadSchoolHolidays() {
        let descriptor = FetchDescriptor<SchoolHolidayPeriod>(
            sortBy: [SortDescriptor(\.startDate)]
        )
        schoolHolidays = (try? modelContext.fetch(descriptor)) ?? []
    }

    // MARK: - Save with Debounce

    func onSettingsChanged() {
        // Latch the picker visibility now (synchronously) so it doesn't
        // disappear during the debounce/save round-trip.
        if workDaysOrHoursChanged {
            hasUnsyncedHistorizedChange = true
        }
        debounceTask?.cancel()
        debounceTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(1))
            guard !Task.isCancelled else { return }
            await self?.saveSettings()
        }
    }

    private func saveSettings() async {
        let descriptor = FetchDescriptor<UserSettings>()
        guard let settings = try? modelContext.fetch(descriptor).first else { return }

        // Stage 2: only attach an effectiveDate when historized fields changed.
        // Otherwise leave nil so the server doesn't write a noise history row.
        let historizedFieldsChanged = workDaysOrHoursChanged
        let effectiveDateForUpload: Date? = historizedFieldsChanged ? effectiveDate : settings.pendingEffectiveDate

        settings.workHoursPerWeek = workHoursPerWeek
        settings.vacationDaysPerYear = vacationDaysPerYear
        settings.workDays = workDays
        settings.bundesland = bundesland
        settings.personalNumber = personalNumber
        settings.updatedAt = Date()
        settings.isPendingSync = true
        settings.isSynced = false
        settings.pendingEffectiveDate = effectiveDateForUpload

        try? modelContext.save()

        // Refresh the "initial" snapshot so subsequent edits compare against
        // the just-saved values.
        initialWorkDays = workDays
        initialWorkHoursPerWeek = workHoursPerWeek

        // Trigger sync
        await syncEngine?.syncAll()

        // After sync completes, re-read the local model. If sync succeeded the
        // pendingEffectiveDate is cleared and we can hide the picker. If sync
        // failed (offline) the marker is still there and the picker remains
        // visible so the user knows the change hasn't shipped yet.
        if let saved = try? modelContext.fetch(FetchDescriptor<UserSettings>()).first {
            hasUnsyncedHistorizedChange = saved.pendingEffectiveDate != nil
        }
    }

    // MARK: - History (Stage 2)

    /// Refreshes the WorkDays / WorkHoursPerWeek timeline from the server.
    func loadSettingsHistory() async {
        guard let client = apiClient else { return }
        isLoadingHistory = true
        historyError = nil
        defer { isLoadingHistory = false }
        do {
            let entries: [UserSettingsHistoryEntryDTO] = try await client.get("/v1/settings/history")
            settingsHistory = entries
        } catch {
            historyError = "Verlauf konnte nicht geladen werden"
        }
    }

    // MARK: - Personal Number (debounced save)

    func onPersonalNumberChanged() {
        personalNumberDebounceTask?.cancel()
        personalNumberDebounceTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(1))
            guard !Task.isCancelled else { return }
            await self?.savePersonalNumber()
        }
    }

    private func savePersonalNumber() async {
        let descriptor = FetchDescriptor<UserSettings>()
        guard let settings = try? modelContext.fetch(descriptor).first else { return }
        settings.personalNumber = personalNumber
        settings.updatedAt = Date()
        try? modelContext.save()
    }

    // MARK: - School Holidays CRUD

    func addSchoolHoliday(name: String, startDate: Date, endDate: Date) {
        let period = SchoolHolidayPeriod(name: name, startDate: startDate, endDate: endDate)
        modelContext.insert(period)
        try? modelContext.save()
        loadSchoolHolidays()
    }

    func updateSchoolHoliday(_ period: SchoolHolidayPeriod, name: String, startDate: Date, endDate: Date) {
        period.name = name
        period.startDate = startDate
        period.endDate = endDate
        try? modelContext.save()
        loadSchoolHolidays()
    }

    func deleteSchoolHoliday(_ period: SchoolHolidayPeriod) {
        modelContext.delete(period)
        try? modelContext.save()
        loadSchoolHolidays()
    }

    // MARK: - Import School Holidays from API

    func importSchoolHolidays() async throws {
        let year = Calendar.current.component(.year, from: Date())
        let stateCode = bundesland.uppercased()
        let url = URL(string: "https://ferien-api.de/api/v1/holidays/\(stateCode)/\(year)")!
        let (data, _) = try await URLSession.shared.data(from: url)

        struct FerienApiHoliday: Decodable {
            let start: String
            let end: String
            let name: String
        }

        let holidays = try JSONDecoder().decode([FerienApiHoliday].self, from: data)

        // Delete all existing school holidays
        let descriptor = FetchDescriptor<SchoolHolidayPeriod>()
        let existing = (try? modelContext.fetch(descriptor)) ?? []
        for period in existing {
            modelContext.delete(period)
        }

        // Insert new ones
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd"

        for h in holidays {
            guard let startDate = dateFormatter.date(from: h.start),
                  let endDate = dateFormatter.date(from: h.end) else { continue }

            let displayName = h.name.split(separator: " ").first.map { String($0).capitalized } ?? h.name

            let period = SchoolHolidayPeriod(
                name: displayName,
                startDate: startDate,
                endDate: endDate
            )
            modelContext.insert(period)
        }

        try? modelContext.save()
        loadSchoolHolidays()
    }

    // MARK: - Logout

    func logout() {
        authManager?.logout()
    }
}
